modules/playlists/src/main/java/org/opencastproject/playlists/PlaylistService.java

```diff
@@ -0,0 +1,398 @@
+    }
+    for (PlaylistEntry entry : playlist.getEntries()) {
+      entry.setPlaylist(playlist);
+    }
+    for (PlaylistAccessControlEntry entry : playlist.getAccessControlEntries()) {
+      entry.setPlaylist(playlist);
+    }
+
+    try {
+      playlist = persistence.updatePlaylist(playlist, securityService.getOrganization().getId());
+      return playlist;
+    } catch (PlaylistDatabaseException e) {
+      throw new IllegalStateException("Could not update playlist from database with id ");
+    }
+  }
+
+  /**
+   * Deletes a playlist from the database
+   * @param playlistId The playlist identifier
+   * @return The removed {@link Playlist}
+   * @throws NotFoundException If no playlist with the given id could be found
+   * @throws IllegalStateException If something went wrong in the database service
+   * @throws UnauthorizedException If the user does not have write access for the playlist
+   */
+  public Playlist remove(String playlistId)
+          throws NotFoundException, IllegalStateException, UnauthorizedException {
+    try {
+      Playlist playlist = persistence.getPlaylist(playlistId);
+      if (!checkPermission(playlist, Permissions.Action.WRITE)) {
+        throw new UnauthorizedException("User does not have write permissions");
+      }
+      playlist = persistence.deletePlaylist(playlist, securityService.getOrganization().getId());
+      return playlist;
+    } catch (PlaylistDatabaseException e) {
+      throw new IllegalStateException("Could not delete playlist from database with id " + playlistId);
+    }
+  }
+
+  /**
+   * Replaces the entries in the playlist with the given entries
+   * @param playlistId identifier of the playlist to modify
+   * @param playlistEntries the new playlist entries
+   * @return {@link Playlist} with the new entries
+   * @throws NotFoundException If no playlist with the given id could be found
+   * @throws IllegalStateException If something went wrong in the database service
+   * @throws UnauthorizedException If the user does not have write access for the playlist
+   */
+  public Playlist updateEntries(String playlistId, List<PlaylistEntry> playlistEntries)
+          throws NotFoundException, IllegalStateException, UnauthorizedException {
+    Playlist playlist;
+    try {
+      playlist = persistence.getPlaylist(playlistId);
+    } catch (PlaylistDatabaseException e) {
+      throw new IllegalStateException(e);
+    }
+    if (!checkPermission(playlist, Permissions.Action.WRITE)) {
+      throw new UnauthorizedException("User does not have write permissions");
+    }
+    playlist.setEntries(playlistEntries);
+
+    try {
+      playlist = persistence.updatePlaylist(playlist, securityService.getOrganization().getId());
+
+      return playlist;
+    } catch (PlaylistDatabaseException e) {
+      throw new IllegalStateException("Could not delete playlist from database with id " + playlistId);
+    }
+  }
```

Comment:
Maybe the function is badly named. It does not actually update specific entries, it replaces all existing entries with the new entries. Furthermore, `playlist.setEntries()` forces the new entries to be part of the playlist. So I am not sure what you are getting at, could you explain in more detail?

Reply:
If it is not intended that you can update an entry, you should ensure that the PlaylistEntry from the update API does not contain the `id` field. e.g., use a DTO that does not contain that field at all.

```
{
  "id" : 999999999
  "contentId":  "ID-about-opencast",
   "type": "EVENT"
}
```

Reply:
If you "update" a playlist with the PlaylistEntry example above, replace the id field with a random number.

You will get an error if you set the ID to the next PK, and it will allow you to update existing entries (intended or not)

Reply:
And what should be the behavior if you want to add a new entry? Send all the old ones (without the ID field) plus the new ones, and they will all get new IDs?

Reply:
> There is a mismatch between the returned ID and the actual data in the database:

Could this be a browser rendering issue? The returned payload looks like it has the right number at least.
![Bildschirmfoto vom 2024-02-27 16-21-05](https://github.com/opencast/opencast/assets/14070005/4b2a9b0b-d1a3-47ff-af7e-a52df8e67b9b)

Reply:
> And what should be the behavior if you want to add a new entry? Send all the old ones (without the ID field) plus the new ones, and they will all get new IDs?

How I intended for this work was that you sent a list with all the entries you want in your playlist to that endpoint. So both existing ones and new ones, and if you want to remove one you just leave it out. The endpoint should also respect ids you set for your entries, so your ids don't change everytime you want to change your entries.

Reply:
Which I guess does allow the user of the endpoint to play mischief and do stuff like

> This allows the user to set arbitrary values for the id field, which results in duplicate key errors if the next autoincremented PK matches.

I guess that can't really be prevented if we stick to just this endpoint? Probably would have to split it into multiple endpoints?

Reply:
> Could this be a browser rendering issue?

It seems so, I never thought that could be a rendering issue :see_no_evil:

Reply:
Something like that should work in the Playlist class:

```java
  public void setEntries(List<PlaylistEntry> entries) {
    for (var entry : entries) {
      if (!this.entries.contains(entry)) {
        entry.setId(0);
      }
      entry.setPlaylist(this);
    }
    this.entries = entries;
  }

  public boolean addEntry(PlaylistEntry entry) {
    notNull(entry, "entry");
    if (this.entries.contains(entry) { //Can be removed if it is a Set
      entries.remove(entry);
    }
    entry.setPlaylist(this);
    return entries.add(entry);
  }
```

That means we need to overwrite the equals and hashCode implementation:

```java
    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
 
        if (!(o instanceof PlaylistEntry))
            return false;
 
        return id != 0 &&
               id.equals(((PlaylistEntry)o).getId());
    }
 
    @Override
    public int hashCode() {
        return getClass().hashCode();
    }
 
```

Furthermore, we can replace the `List<PlayEntry>` with `Set<PlayEntry>`, which would protect us against duplicate entries with the same ID.

Or do something equivalent in the PlaylistService methods.

Reply:
Tried to apply what you suggested.

Stuck with `List`, since ordering is important to us, and it seems using something like `LinkedHashSet` would get overly complicated with JPA and JAXB.

Also `POST /{id}/entries.json` still does not return the correct playlist, but I have not been able to find out why that is.

Reply:
Returning the result of the merge should solve that.